tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java

```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
 		return separator;
 	}
 
-	private void _makeActions(ISelectionProvider provider) {
+	private void _makeActions(IMenuManager menu, ISelectionProvider provider) {
+		ISelection selection = provider.getSelection();
+
+		if (selection == null) {
+			return;
+		}
+
 		_watchAction = new WatchWorkspaceModulesAction(provider);
 		_stopAction = new StopWorkspaceModulesAction(provider);
+
+		if (selection instanceof TreeSelection) {
+			TreeSelection watchSelection = (TreeSelection)selection;
+
+			List<?> selectProjects = watchSelection.toList();
+
+			if (ListUtil.hasMultipleValues(selectProjects)) {
+				_watchAction.setEnabled(true);
+				_stopAction.setEnabled(true);
+
+				menu.add(_watchAction);
+				menu.add(_stopAction);
+
+				return;
+			}
+
+			Object selectedProject = watchSelection.getFirstElement();
+
+			IWorkspaceProject workspaceProject = LiferayCore.create(IWorkspaceProject.class, selectedProject);
+
+			if (workspaceProject != null) {
+				IProject project = workspaceProject.getProject();
+
+				Set<IProject> watchingProject = workspaceProject.watching();
+
+				String jobName = project.getName() + ":watch";
+
+				IJobManager jobManager = Job.getJobManager();
+
+				Job[] jobs = jobManager.find(jobName);
+
+				if (ListUtil.isNotEmpty(jobs) && ListUtil.contains(watchingProject, project)) {
+					_watchAction.setEnabled(false);
+					_stopAction.setEnabled(true);
+				}
+				else {
+					_watchAction.setEnabled(true);
+					_stopAction.setEnabled(false);
+				}
+
+				menu.add(_watchAction);
+				menu.add(_stopAction);
+			}
+			else {
+				TreePath[] selectionPaths = watchSelection.getPathsFor(selectedProject);
+
+				Object parentObject = Stream.of(
+					selectionPaths
+				).filter(
+					selectionPath -> selectedProject.equals(selectionPath.getLastSegment())
+				).map(
+					selectPath -> {
+						TreePath parentPath = selectPath.getParentPath();
+
+						return parentPath.getLastSegment();
+					}
+				).findFirst(
+				).get();
+
+				IWorkspaceProject parentWorkspaceProject = LiferayCore.create(IWorkspaceProject.class, parentObject);
```

Comment:
This logic below looks very much like logic above.  Can this be simplified or extracted?

Reply:
I tried to simplified it, but the two part has some different logic. The first part, I can only focus on workspace project, the second, I have to focus on them same time.